[CAMEL-15341] Create camel-azure-storage-datalake to serve Azure DataLakes Gen2

aryankoul opened the pull request:
New component that handles operations for azure datalake storage using  the Azure SDK for java.
At present, It supports operations related to files and filesystems in datalake.

oscerd commented on components/camel-azure-storage-datalake/pom.xml:
```diff
@@ -0,0 +1,89 @@
+        <dependency>
+            <groupId>com.azure</groupId>
+            <artifactId>azure-storage-file-datalake</artifactId>
+            <version>12.2.0</version>
```
This needs to be a property placeholder in parent/pom

oscerd commented on components/camel-azure-storage-datalake/pom.xml:
```diff
@@ -0,0 +1,89 @@
+        <dependency>
+            <groupId>com.azure</groupId>
+            <artifactId>azure-identity</artifactId>
+            <version>1.1.3</version>
```
Same

oscerd reviewed (changes requested):
Just two minor findings. Let's wait for @omarsmak

omarsmak commented on components/camel-azure-storage-datalake/src/main/java/org/apache/camel/component/azure/storage/datalake/DataLakeComponent.java:
```diff
@@ -0,0 +1,126 @@
+
+    private void validateConfiguration(final DataLakeConfiguration config) {
+        if (config.getServiceClient() == null && config.getClientSecretCredential() == null) {
+            throw new IllegalArgumentException("cliet or credentials must be specified");
```
typo here

omarsmak commented on components/camel-azure-storage-datalake/src/main/java/org/apache/camel/component/azure/storage/datalake/DataLakeComponent.java:
```diff
@@ -0,0 +1,126 @@
+        }
+    }
+
+    private void setClientFromRegistry(final DataLakeConfiguration configuration) {
```
Actually we no longer need this, we can now autowire the client by adding `@Metadata(autowired = true)` to the serviceClient configuration and remove this. An example [here](https://github.com/apache/camel/commit/bc8fa176a66bfab4b5385305c626fddc01d29c2b#diff-fa0be9a42d8b24af5d483c6e806968d302c829087bcf363ce5fc2e67559ca863)

omarsmak commented on components/camel-azure-storage-datalake/src/main/java/org/apache/camel/component/azure/storage/datalake/DataLakeEndpoint.java:
```diff
@@ -0,0 +1,97 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@UriEndpoint(firstVersion = "3.5.0", scheme = "azure-storage-datalake", title = "Azure storage datalake service",
```
the first version should be 3.7.0

omarsmak commented on components/camel-azure-storage-datalake/src/main/java/org/apache/camel/component/azure/storage/datalake/operations/DataLakeDirectoryOperations.java:
```diff
@@ -0,0 +1,33 @@
+import org.apache.camel.component.azure.storage.datalake.client.DataLakeDirectoryClientWrapper;
+import org.apache.camel.util.ObjectHelper;
+
+public class DataLakeDirectoryOperations {
```
Is there any good use for this class?

aryankoul replied:
So, there are some operations on directory as well in datalake gen2. I am gonna add those in here.

omarsmak reviewed (changes requested):
Thank you @aryankoul for the PR! At a first glance, the PR looks good with some minor requests. However, one big request I'd like is to add more comprehensive tests that cover both unit tests and integration tests. I think you can still follow [camel storage blob for the test](https://github.com/apache/camel/tree/c2ee8e278e44a7e65c701dbd3502991f9fcaa3ad/components/camel-azure-storage-blob/src/test/java/org/apache/camel/component/azure/storage/blob) strategy for both unit and integration as both components pretty similar

davsclaus commented:
Is there any update on this? As we are closing down for 3.7 release so this needs to be done soon, otherwise it will miss the release

oscerd commented:
I think this is for 3.8 .. we are cutting the release already

aryankoul commented:
> Is there any update on this? As we are closing down for 3.7 release so this needs to be done soon, otherwise it will miss the release

I'll need some time for adding proper unit and integration tests.

oscerd commented:
Can you please rebase and solve conflicts? Thanks

aryankoul commented:
> Can you please rebase and solve conflicts? Thanks

Sure.

oscerd commented:
I would like to have feedback from @omarsmak too. By the way the branch is a bit a mess and it cannot be rebased. Can you please amend the situation?

omarsmak commented on components/camel-azure-storage-datalake/src/main/java/org/apache/camel/component/azure/storage/datalake/DataLakeConstants.java:
```diff
@@ -0,0 +1,86 @@
+    public static final String PUBLIC_ACCESS_TYPE = HEADER_PREFIX + "PublicAccessType";
+    public static final String RAW_HTTP_HEADERS = HEADER_PREFIX + "RawHttpHeaders";
+    public static final String DATALAKE_REQUEST_CONDITION = HEADER_PREFIX + "RequestCondition";
+    public static final String LIST_PATH_OPTIONS = HEADER_PREFIX + "ListPathOptions";
```
This constant seems to be used, can you please check?

omarsmak commented on components/camel-azure-storage-datalake/src/main/java/org/apache/camel/component/azure/storage/datalake/DataLakeConsumer.java:
```diff
@@ -0,0 +1,166 @@
+    protected int poll() throws Exception {
+        final String fileSystemName = getEndpoint().getConfiguration().getFileSystemName();
+        final String fileName = getEndpoint().getConfiguration().getFileName();
+        final String directoryName = getEndpoint().getConfiguration().getDirectoryName();
```
`directoryName` is not used here, any purpose?

omarsmak commented on components/camel-azure-storage-datalake/src/main/java/org/apache/camel/component/azure/storage/datalake/DataLakeProducer.java:
```diff
@@ -0,0 +1,169 @@
+        return new DataLakeFileSystemOperations(configuration, clientWrapper);
+    }
+
+    private DataLakeDirectoryOperations getDirectoryOperations(final Exchange exchange) {
```
I see this operation is not used

omarsmak commented on catalog/camel-catalog/src/generated/resources/org/apache/camel/catalog/docs/azure-storage-datalake-component.adoc:
```diff
@@ -0,0 +1,119 @@
+[[azure-storage-datalake-component]]
```
The documentation is laking a lot of info, please add info about the component, examples, usage, info about the headers that users can use in the producer and expect in the consumer,  any features that you implemented and you want users to try out ... etc. Ex:  https://camel.apache.org/components/latest/azure-storage-blob-component.html

omarsmak commented on test-infra/camel-test-infra-azure-storage-datalake/src/test/java/org/apache/camel/test/infra/azure/storage/datalake/clients/AzureStorageDataLakeClientUtils.java:
```diff
@@ -0,0 +1,48 @@
+        String accountKey = System.getProperty(AzureConfigs.ACCOUNT_KEY);
+        StorageSharedKeyCredential credential = new StorageSharedKeyCredential(accountName, accountKey);
+
+        String host = System.getProperty(AzureConfigs.HOST);
```
I see `HOST` and `PORT` is not used. Can you remove these since Azurite is not yet supporting `Datalake`

omarsmak reviewed (changes requested):
@aryankoul thanks for the modifications, things it seems to be in a good direction, however there are couple of things need to change as I commented. Also, this branch it seems to be compiled with 3.7 not with camel 3.8. Can you please fix the branch and have it compiled with 3.8?
Also, if you can add a comment on how to run the integration tests would be appreciated so I can run it from my end. Thanks

aryankoul commented on components/camel-azure-storage-datalake/src/main/java/org/apache/camel/component/azure/storage/datalake/DataLakeEndpoint.java:
```diff
@@ -0,0 +1,97 @@
+    protected void doStart() throws Exception {
+        super.doStart();
+
+        if (configuration.getServiceClient() != null) {
```
Can you please elaborate a bit?

omarsmak replied:
Actually what you did here is fine, I forgot that will help to bind the client into the endpoint params, hence please ignore this comment :)

omarsmak commented:
hey @aryankoul, any progress on this PR?

aryankoul commented:
> hey @aryankoul, any progress on this PR?

Hi @omarsmak , i have pushed new changes. please review the same.

omarsmak reviewed (approved):
LGTM! Thank you @aryankoul ! 
Once this is merged, the next step to create the Spring boot starter in camel-spring-boot and if you have time, would be great if you can add an example demonstrate the component in the camel-examples repo, you can re-use [this](https://github.com/apache/camel-examples/tree/master/examples/camel-example-debezium-eventhubs-blob) example for datalake component. 

Let's see if others have any other comments cc @davsclaus @oscerd

oscerd reviewed (approved):
LGTM, I think we'll need a bundle for the karaf side. We need also to resolve the conflicts.

aryankoul commented:
> LGTM! Thank you @aryankoul !
> Once this is merged, the next step to create the Spring boot starter in camel-spring-boot and if you have time, would be great if you can add an example demonstrate the component in the camel-examples repo, you can re-use [this](https://github.com/apache/camel-examples/tree/master/examples/camel-example-debezium-eventhubs-blob) example for datalake component.
> 
> Let's see if others have any other comments cc @davsclaus @oscerd
> 
> P/S: there are some conflicts, can you please resolve them? thanks

Thanks @omarsmak!
Sure, i'll look into the camel-examples repo.

oscerd commented:
@aryankoul can you rebase please?

omarsmak commented:
Can we merge this PR?

oscerd commented:
It cannot be rebased. But I'll try to merge it today or tomorrow

omarsmak commented:
> It cannot be rebased. But I'll try to merge it today or tomorrow

Thanks a lot!

oscerd commented:
Merged but tests are failing 

[ERROR] Errors: 
[ERROR] org.apache.camel.component.azure.storage.datalake.integration.DataLakeConsumerTest.null
[ERROR]   Run 1: DataLakeConsumerTest>BaseIT.initProperties:80 » NullPointer 'accountName' cann...
[ERROR]   Run 2: DataLakeConsumerTest>BaseIT.initProperties:80 » NullPointer 'accountName' cann...
[ERROR]   Run 3: DataLakeConsumerTest>BaseIT.initProperties:80 » NullPointer 'accountName' cann...
[INFO] 
[ERROR] org.apache.camel.component.azure.storage.datalake.integration.DataLakeFileOperationTest.null
[ERROR]   Run 1: DataLakeFileOperationTest>BaseIT.initProperties:80 » NullPointer 'accountName'...
[ERROR]   Run 2: DataLakeFileOperationTest>BaseIT.initProperties:80 » NullPointer 'accountName'...
[ERROR]   Run 3: DataLakeFileOperationTest>BaseIT.initProperties:80 » NullPointer 'accountName'...
[INFO] 
[ERROR] org.apache.camel.component.azure.storage.datalake.integration.DataLakeFileSystemOperationTest.null
[ERROR]   Run 1: DataLakeFileSystemOperationTest>BaseIT.initProperties:80 » NullPointer 'accoun...
[ERROR]   Run 2: DataLakeFileSystemOperationTest>BaseIT.initProperties:80 » NullPointer 'accoun...
[ERROR]   Run 3: DataLakeFileSystemOperationTest>BaseIT.initProperties:80 » NullPointer 'accoun...

oscerd commented:
For the moment it is in a local branch on my machine, correctly rebased. Please fix the tests.

aryankoul commented:
> For the moment it is in a local branch on my machine, correctly rebased. Please fix the tests.

@oscerd i have fixed the tests, please check now.

oscerd commented:
If you run the mvn install on this component the tests will run anyway and fail. No matter the profile.

oscerd commented:
If you have docker service running, they will run with a pure mvn install.

oscerd commented:
Lets try to find a solution, because we would like to release 3.8 soon.

aryankoul commented:
> If you run the mvn install on this component the tests will run anyway and fail. No matter the profile.

AFAIK , only those files are included automatically for test which have "Test" as prefix or suffix. Thats why i changed the names of the test files that were failing due to lack of access key,  to have prefix "IT" instead of "Test", so that they do not run automatically in mvn commands. I have added a profile to run them explicitly.

oscerd commented:
Rebase and run a full build mvn install -DskipTests.. they will run if you have docker service running. You have to deactivate the profile by default, ignore the test, remove the tests or find a different approach.

oscerd commented:
I mean run the full build and then run mvn install only on the component. They will run

oscerd commented:
Merged with some modification. Thanks
